essentials/src/api/api_client.rs

```diff
@@ -15,66 +15,249 @@
 // along with polkadot-introspector.  If not, see <http://www.gnu.org/licenses/>.
 //
 
-use crate::types::{BlockNumber, Header, H256};
+use crate::{
+	metadata::polkadot,
+	types::{AccountId32, BlockNumber, Header, InherentData, SessionKeys, SubxtHrmpChannel, Timestamp, H256},
+};
+use dyn_clone::DynClone;
+use std::collections::BTreeMap;
 use subxt::{
 	backend::{
 		legacy::{rpc_methods::NumberOrHex, LegacyRpcMethods},
 		rpc::RpcClient,
 		StreamOf,
 	},
-	blocks::{BlockRef, BlocksClient},
-	events::EventsClient,
+	blocks::{Block, BlockRef, BlocksClient},
+	client::{LightClient, OnlineClientT},
+	dynamic::Value,
+	events::{Events, EventsClient},
 	storage::StorageClient,
 	OnlineClient, PolkadotConfig,
 };
 
+pub type HeaderStream = StreamOf<Result<(Header, BlockRef<H256>), subxt::Error>>;
+
+#[async_trait::async_trait]
+pub trait ApiClientT: DynClone + Send + Sync {
```

Comment:
Would we ever need to have two different implementations of this trait to easily switch between them ? If answer is no I think we don't really need a trait at the moment.

Reply:
I think we should keep both of them. The Light Client is slow, it takes some time to synchronize with the network. This can be annoying when we are looking for a historical data or need a quick run to get an immediate result.